com.unity.render-pipelines.core/Editor/Lighting/GUIFramework/Control.cs

```diff
@@ -0,0 +1,279 @@
+using System;
+using UnityEngine;
+
+namespace UnityEditor
+{
+    /// <summary>
+    /// An interface that represents a UI control.
+    /// </summary>
+    public abstract class Control
+    {
+        private string m_Name;
+        private int m_NameHashCode;
+        private int m_ID;
+        private LayoutData m_LayoutData;
+        private int m_ActionID = -1;
+        private LayoutData m_HotLayoutData;
+        private bool m_Enabled;
+
+        /// <summary>
+        /// The name of the control.
+        /// </summary>
+        public string name
+        {
+            get { return m_Name; }
+        }
+
+        /// <summary>
+        /// The control ID. The GUI uses this to identify the control.
+        /// </summary>
+        public int ID
+        {
+            get { return m_ID; }
+        }
+
+        /// <summary>
+        /// Action ID
+        /// </summary>
+        public int actionID
+        {
+            get { return m_ActionID; }
+        }
+
+        /// <summary>
+        /// Layout Data
+        /// </summary>
+        public LayoutData layoutData
+        {
+            get { return m_LayoutData; }
+            set { m_LayoutData = value; }
+        }
+
+        /// <summary>
+        /// Hot layout Data
+        /// </summary>
+        public LayoutData hotLayoutData
+        {
+            get { return m_HotLayoutData; }
+        }
+
+        /// <summary>
+        /// Initializes and returns an instance of Control
+        /// </summary>
+        /// <param name="name">The name of the control</param>
+        public Control(string name)
+        {
+            m_Name = name;
+            m_NameHashCode = name.GetHashCode();
+        }
+
+        /// <summary>
+        /// Gets the control from the guiState.
+        /// </summary>
+        /// <param name="guiState">The current state of the custom editor.</param>
+        public void GetControl(IGUIState guiState)
+        {
+            if (guiState.eventType == EventType.Layout)
+                m_Enabled = GetEnabled(guiState);
+
+            m_ID = -1;
+
+            if (m_Enabled)
+                m_ID = guiState.GetControlID(m_NameHashCode, FocusType.Passive);
+        }
+
+        internal void SetActionID(int actionID)
+        {
+            m_ActionID = actionID;
+            m_HotLayoutData = m_LayoutData;
+        }
+
+        /// <summary>
+        /// Begins the layout for this control. A call to EndLayout must always follow a call to this function.
+        /// </summary>
+        /// <param name="guiState">The current state of the custom editor.</param>
+        public void BeginLayout(IGUIState guiState)
```

Comment:
I am not a big fan of BeginLayout, Layout EndLayout that are public and must be called in this order. This is error prone for user trying to use this and there is no feedback stating that there is for instance no EndLayout.

A couple Begin/End could be written as a scope. (use struct to prevent GC). This should prevent any missuse regarding Begin/End balance.

And as Layout is not overrideable, I wonder why this is not one big function that call BeginLayout, Layout and then EndLayout in that order. They are only used by GUISystem.